[RFC] ASoC: SOF: topology: use set_get_data in process load

juimonen opened the pull request:
Currently when loading sof process components there's a check if binary
control data is associated with it. If found the data is extracted to be
part of component loading and initialization. If binary data exceeds the
ipc max size, loading fails with error as large message support is only
implemented in set_get_data method. So make the process loading use
set_get_data to enable large parameters in component initialization.

Signed-off-by: Jaska Uimonen <jaska.uimonen@intel.com>

juimonen commented:
I'm not sure if there's restrictions why we could not send params during component loading...
corresponding FW PR:
https://github.com/thesofproject/sof/pull/1408

plbossart commented on include/uapi/sound/sof/abi.h:
```diff
@@ -26,7 +26,7 @@
 /* SOF ABI version major, minor and patch numbers */
 #define SOF_ABI_MAJOR 3
-#define SOF_ABI_MINOR 5
+#define SOF_ABI_MINOR 6
```
is this needed?

juimonen replied:
no

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1738,63 +1738,18 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 	size_t ipc_size, ipc_data_size = 0;
-	int ret, i, offset = 0;
+	int ret, i;
+	int ipc_cmd, ctrl_type;
```
reverse xmas tree and one declaration per line if possible.

juimonen replied:
ok

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1738,63 +1738,18 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 	ipc_size = sizeof(struct sof_ipc_comp_process) +
 		le32_to_cpu(private->size) +
-		ipc_data_size;
+	  ipc_data_size;
```
alignment?

juimonen replied:
ok

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1738,63 +1738,18 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 	ipc_size = sizeof(struct sof_ipc_comp_process) +
 		le32_to_cpu(private->size) +
-		ipc_data_size;
+	  ipc_data_size;
```
I don't get at all what this simplification does, we now have ipc_data_size = 0?

juimonen replied:
removed totally.

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1819,6 +1774,17 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 
 	sof_dbg_comp_config(scomp, &process->config);
 
+	process->size = ipc_data_size;
```
ipc_data_size is still zero here?

juimonen replied:
removed.

singalsu commented:
@cujomalainey This work fixes the embedding of long EQ filter coefficients set into topology file.

cujomalainey commented:
Much appreciated, once we have this on a chromebook I will finish up testing on the GUI then send a PR.

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -1819,6 +1774,17 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
+
 	/*
 	 * found private data in control, so copy it.
 	 * get possible component controls - get size of all pdata,
```
@juimonen the comment is no longer valid.

juimonen replied:
ok

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -1856,17 +1824,39 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
+		case SOF_CTRL_CMD_SWITCH:
+			ipc_cmd = SOF_IPC_COMP_SET_VALUE;
+			ctrl_type = SOF_CTRL_TYPE_VALUE_CHAN_SET;
+			break;
```
@juimonen I'm not sure about this. We cant pass any private data in the kcontrol as a volume set command. Either there has to be a sanity check to make sure it is in the volume table or we remove it. I think it is unnecessary to pass any other data that the binary control data.

juimonen replied:
ok

ranj063 commented:
@juimonen Did you test this change with the kwd topology. The selector and the KPB components expect binary data to be attached to the process ipc struct payload. This change will break that requirement

juimonen commented:
@ranj063 I did not test it with KWD. But KWD should also support large data and message sending? I mean the whole large data was done for KWD...?

juimonen commented:
@plbossart @ranj063 thanks for reviewing. it is me leaving friday evening and quickly pushing something for others to see... the abi change is pointless, I didn't notice it, I only need it for my local testing. ipc_data_size should be also removed. 

I only wanted to get first high level comments is there some conceptual problem of not using set_get_data in component creation. Without fw side PR this change will actually oops the DSP...

So fw/dsp implicitly assumes that fw side component should have valid pipeline pointer when these messages are sent. Neither side is checking it. if I understand correctly the pipeline pointer is valid only when pipeline is "complete".  But I don't get why we should restrict of sending controls only when pipeline is ready...

juimonen commented:
@plbossart with this PR I want to enable large messages in component creation.

ranj063 commented:
@juimonen we talked about sending default configuration parameters along as binary data as ipc payload when new components are created. So for components like selector or the kpb, we should think about adding the config data to the widget instead if having a binary control.

juimonen commented:
@ranj063 my point here is: if you use controls to have binary data, you have to be prepare in component that someone is actually using the control, so not just in creation phase...?

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1812,61 +1765,83 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 	for (i = 0; i < widget->num_kcontrols; i++) {
 		kc = &widget->kcontrol_news[i];
 
+		/* we only care about bytes control */
```
why? Isn't that a restriction compared to the previous solution?

juimonen replied:
@plbossart currently we are using component related control's binary blob to convey binary parameters to "process" component. and yeah, the name was not my idea...

As @ranj063 pointed out in another comment, why to even try to send other control types like enums or volume data at this point because it doens't make sense? Hence, I only try to find binary and "continue" with other types, if someone has tried to push that from topology, for some strange reason.

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1728,73 +1728,26 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 	ipc_size = sizeof(struct sof_ipc_comp_process) +
-		le32_to_cpu(private->size) +
-		ipc_data_size;
+		le32_to_cpu(private->size);
```
that one I just can't figure out? Looks to me like things may work for binary controls but will no longer work for other types of controls.
Confused.

juimonen replied:
@plbossart see above. but I'm not totally sure about this, it is very easy to put back the parsing for other types of data. @ranj063 what do you think?

plbossart replied:
I don't see what's different between the various types of kcontrols. If you can set them in topology then it should be the same handling for all.

juimonen replied:
@plbossart, ok fine by me, i'll add them back

lyakh commented:
I'm probably more confused by this than anybody else, trying to review this. Isn't this commit trying to do several things? Changing the logic and increasing the maximum size of data, that can be processed? Shouldn't at least they be separated?

juimonen commented:
@lyakh so previously data was dig out from component related control and copied to component ipc struct. Now as large messages is only supported with set_get_data, this previous method can't send big parameters as part of the component creation struct. This PR is just removing this copy, so component is created with empty binary data and after it is created I'm sending the private binary data with set_get_data (which support large messages). Instead of previously only sending 1 ipc in creation, I'm sending 2 ipc (component creation then the data). 

I think this is quite simple. The only thing to "understand" that we might have component related private data in its related control. This was originally done because it is really nasty to include both tokens and binary blob to component in topology, so it is nasty from kernel point of view when parsing this. There needed to be all kinds of hard coded offsets etc. So we decided to add the possible binary blob to related control. Despite this, if you remember, I was specifically told to implement the large messages only to control set/get data, so not universally (my first PR was for universal IPC and it was not accepted) -> if we try to create component bigger than max ipc size it will fail -> let's use the set_get_data in creation so we can send bigger params.

In the end, the most "correct" way of doing this would be to 1) enable adding binary blob together with tokens to component topology and parsing it nicely. Then 2) adding large message support for component creation. But this is not as easy and straightforward as this PR...

keyonjie commented:
This looks legit to me, though it will break today's KWD, as we don't have set_get_data support for kpb and selector ATM.
@mrajwa @lbetlej we might need add set_get_data support there(similar to detector component).

juimonen commented:
@keyonjie ok, I see detector and selector support this but not kpb..

juimonen commented:
@plbossart @ranj063 @keyonjie updated. This version supports legacy components, so it is backwards compatible. Data fitting to single ipc is sent with "old method". There's also a 2nd commit which refactors the process loading function. It it looks fine it can be squashed to the first one. Just to make review easier...

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1794,11 +1799,17 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 		return -ENOMEM;
 
-	/* configure iir IPC message */
+	/* configure process IPC message */
```
nit-pick: should be separate patch?

juimonen replied:
ok

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1858,13 +1873,77 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
+	for (i = 0; i < widget->num_kcontrols; i++) {
+		kc = &widget->kcontrol_news[i];
+
+		/* we only care about bytes control */
```
you should elaborate on this, the comment is just too short to understand what you explained in the previous review (it doesn't make sense to send the other types through topology).

juimonen replied:
actually this is now just wrong, will remove.

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1858,13 +1873,77 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
+		if (!scontrol->control_data->data)
+			continue;
+
+		/* make sure data is valid - data can be updated at runtime */
```
is the comment correct? I thought we are in the part where we send stuff from topology, so what can be updated?

juimonen replied:
this is an old comment so this is mystery to me also... so I know that data can be updated runtime but it doesn't too much connect to "validity". I'll fix it in separate commit.

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1821,59 +1858,30 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 			       le32_to_cpu(private->size));
 	if (ret != 0) {
-		dev_err(sdev->dev, "error: parse process.cfg tokens failed %d\n",
+		dev_err(sdev->dev, "error: parse process tokens failed %d\n",
```
separate patch?

juimonen replied:
well, yeah I'm just fixing things what check-patch wants me to do, I should then do intermediate commits without hooks. These are old stuff that checkpatch doesn't like. Anyway I can put all comment changes to separate patch.

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1749,50 +1825,11 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
-	 * get possible component controls - get size of all pdata,
-	 * then memcpy with headers
-	 */
+	/* get possible component controls and get size of all pdata */
```
separate patch? is this change needed?

juimonen replied:
ok, removed the change

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1719,6 +1719,85 @@ static int sof_widget_load_siggen(struct snd_soc_component *scomp, int index,
 	return ret;
 }
 
+static int sof_get_control_data(struct snd_sof_dev *sdev,
```
@juimonen general comment on this comment:
On one hand I appreciate the logic of trying to make the change then refactor
On the other hand, this looks like tick-tock patches where you change what you just introduced.
Maybe make sure that this a block change and avoid additional changes such as changed comments or changed sequences, etc?

juimonen replied:
@plbossart I tried to remove my changes to all non-related comments even they would be originally erroneous. That can be addressed in another pr.

ranj063 commented:
@juimonen sorry for the delay. I like the refactoring part but I have a suggestion to make it a bit more intuitive. We still call the sof_get_control_data() method 3 times. We should only call it once and gather all the information in one go. Basically, it should return the total size of the data for all controls as the return value. Additionally, it should create a pdata and scontrol array of size num_controls. We could possibly create a struct like this (for example):

struct widget_data {
void **pdata;
struct snd_sof_control **control;
int ipc_cmd;
int cmd_type;
};

and pass it to the sof_get_control_data() and populate everything we need. Once this is done, you can determine the course of action based on the returned size.

juimonen commented:
@ranj063 heh yes, was thinking on the same lines, just little bit worried of using dynamic arrays in kernel code (don't have so much experience on this) and then thinking is it a overkill for removing 3 iterations? But agree that would be the cleanest solution...

ranj063 commented:
> then thinking is it a overkill for removing 3 iterations? But agree that would be the cleanest solution...

@juimonen I guess the problem right now is that no one other than you and me understand why this is being done. And calling the same function thrice is only going to confuse people more. Whereas if we do it once, we can add proper comments to say that we're calculating the total size, copying the data and saving the control data for IPC's, it will make things clearer.

juimonen commented:
@ranj063 yes sure, i'll try to do it like that.

juimonen commented:
@plbossart @ranj063 updated to Ranjani's comments. I tried to leave any extra touching of old comments away...not sure it totally succeeded. As said I think these patches should be anyway squashed...

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -1858,13 +1873,76 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 				 ipc_size, r, sizeof(*r));
-	if (ret >= 0)
+
+	if (ret >= 0 && ipc_data_size)
```
@juimonen we should also return is ret < 0 because that means the ipc message has failed. so I think a check for if (ipc_data_size) should suffice here.

juimonen replied:
@ranj063 hmm ok this is dodgy... I'll refine it and remove the goto's to make it clear.

juimonen replied:
ok

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -1777,22 +1760,81 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
+		ipc_data_size = sof_get_control_data(sdev, widget, wdata);
 
-		ipc_data_size += pdata->size;
+		if (ipc_data_size <= 0) {
```
@juimonen is it also an error is ipc_data_size is 0? if the process comp has no controls, this will be 0 isnt it?

juimonen replied:
@ranj063 I think we should not be here if there's no controls... ? this is under:
if (widget->num_kcontrols) {

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -1828,122 +1872,45 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 			goto err;
 		}
 	}
 
```
@juimonen there's an error in freeing process here. kfree(process) would be called even if there's no error which will results in swidget->private being NULL and this will cause the component set up fail upon resume.

juimonen replied:
@ranj063 yes good catch, I'll remove the goto's because they make this ugly. I could do this with couple of goto's but that start's to be too much

plbossart commented:
@juimonen can you address comments and mark them as resolved when applicable. It's difficult to review when we don't know if the comments are stale or still apply. Thanks!

juimonen commented:
@ranj063 you have time still to take a look? thanks!

ranj063 commented:
@juimonen looks good now but I think these should be squashed or the refactor done first.

juimonen commented:
@ranj063 yes I made the separate re-factor for @plbossart (cause I know he likes them separately). @plbossart if you're good, I'll squash these commits together?

plbossart commented:
I can't look into this in details so what @ranj063 says

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1822,51 +1879,49 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
+
+	/* we sent the data in single message so return */
+	if (ipc_data_size) {
+		kfree(wdata);
```
is there a missing kfree(process) here? 
Or is this intentional, and might be worth documenting.

ranj063 replied:
@plbossart this is intentional. process is saved as swidget->private for loading the comp upon resuming from PM suspend

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1822,51 +1879,49 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
+						    true);
+		if (ret != 0) {
+			dev_err(sdev->dev, "error: send control failed\n");
+			kfree(process);
```
and this looks wrong, you might free the same thing multiple times in the loop - that may be harmless but it is a telltale sign of memory allocation issues. if there a missing return?

ranj063 replied:
@plbossart good catch. we need a break after kfree(process). @juimonen is out for the next week or so. I will update the PR on his behalf.

juimonen replied:
ok

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1822,51 +1879,49 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
+		}
+	}
+
+	kfree(wdata);
```
and here process is not freed either

juimonen commented:
@plbossart so "wdata" is local dynamic variable which should be freed after errors and end of function. I have no idea how many controls this process has and I don't want to put there static array. 

"process" is not local so it should be freed only after error. 

As said I changed this to get rid of ugly goto's and the missing break in the loop you found is a real issue and fixed now by @ranj063

juimonen commented:
@ranj063 @plbossart updated the break also here.

juimonen commented:
SOFCI TEST

ranj063 commented:
@juimonen could you please rebase ?

juimonen commented:
@ranj063 rebased

ranj063 commented:
@juimonen still shows "This branch is out-of-date with the base branch"?

juimonen commented:
@ranj063 oh accidentally pushed the branch to sof repo than my own. should be now rebased, will delete the branch from origin...

ranj063 commented:
@juimonen unfortunately @plbossart has to approve this PR before it can be merged because he had requested changes.

juimonen commented:
@ranj063 no worries and thanks for making the extra PR effort.

juimonen commented:
@ranj063 @plbossart rebased. I've no idea why some pm_status check is failing...

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1795,31 +1783,98 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 		return -ENOMEM;
+	}
 
 	/* configure iir IPC message */
```
could we please fixup this comment in a follow-up patch? Thanks!
